internal/service/instance/nginx.go

```diff
@@ -0,0 +1,203 @@
+/**
+ * Copyright (c) F5, Inc.
+ *
+ * This source code is licensed under the Apache License, Version 2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package instance
+
+import (
+	"bufio"
+	"bytes"
+	"log/slog"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/nginx/agent/v3/api/grpc/instances"
+	process "github.com/nginx/agent/v3/internal/datasource/nginx"
+	"github.com/nginx/agent/v3/internal/datasource/os/exec"
+	"github.com/nginx/agent/v3/internal/model"
+	"github.com/nginx/agent/v3/internal/uuid"
+)
+
+var (
+	re     = regexp.MustCompile(`(?P<name>\S+)/(?P<version>\S+)`)
+	plusre = regexp.MustCompile(`(?P<name>\S+)/(?P<version>\S+).\((?P<plus>\S+plus\S+)\)`)
+)
+
+type Info struct {
+	Version       string
+	PlusVersion   string
+	Prefix        string
+	ConfPath      string
+	Cfgf          map[string]interface{}
+	ConfigureArgs []string
+}
+
+type Nginx struct {
+	executer exec.ExecInterface
+}
+
+type NginxParameters struct {
+	executer exec.ExecInterface
+}
+
+func NewNginx(parameters NginxParameters) *Nginx {
+	if parameters.executer == nil {
+		parameters.executer = &exec.Exec{}
+	}
+	return &Nginx{
+		executer: parameters.executer,
+	}
+}
+
+func (n *Nginx) GetInstances(processes []*model.Process) ([]*instances.Instance, error) {
+	var processList []*instances.Instance
+
+	nginxProcesses := make(map[int32]*model.Process)
+	for _, p := range processes {
+		if isNginxProcess(p.Name, p.Cmd) {
+			nginxProcesses[p.Pid] = p
+		}
+	}
+
+	for _, nginxProcess := range nginxProcesses {
+		_, ok := nginxProcesses[nginxProcess.Ppid]
```

Comment:
We are iterating through `nginxProcesses`, so why are we checking whether `nginxProcess` is in `nginxProcesses`?

Reply:
we are checking that the parent of the nginx process is in the map or not to determine if its a worker or master process

Reply:
Ah, sorry I missed the extra 'P' of `Ppid`. Was super confused there 😅